File: simcoe/src/main/java/com/prolificinteractive/simcoe/Tracker.java
package com.prolificinteractive.simcoe;

import com.prolificinteractive.simcoe.trackers.AnalyticsTracking;
import java.util.ArrayList;
import java.util.List;

/**
 * A recorder for SimcoeEvents.
 */
public final class Tracker {

  private static final String ANALYTICS_PROVIDER_NAME = "Analytics";
  private static final String WRITE_OUT_FORMAT = "[%s] %s";
  private static final String ANALYTICS_ERROR = "** ANALYTICS ERROR **";

  private final List<Event> events = new ArrayList<>();
  private final List<Output> outputSources;

  /**
   * The output options.
   */
  OutputOptions outputOption = OutputOptions.VERBOSE;

  /**
   * The error options.
   */
  ErrorHandlingOption errorOption = ErrorHandlingOption.DEFAULT;

  /**
   * Creates a new tracker instance.
   *
   * @param outputSources The source to use for general output. If no output sources are provided,
   * the default is set as the standard output console.
   */
  public Tracker(final List<Output> outputSources) {
    if (outputSources != null) {
      this.outputSources = outputSources;
    } else {
      this.outputSources = new ArrayList<>();
      this.outputSources.add(new ConsoleOutput());
    }
  }

  private void handleError(final String message) throws
      Exception {
    switch (errorOption) {
      case DEFAULT:
        writeOut(ANALYTICS_ERROR, message);
        break;
      case STRICT:
        throw new Exception(message);
      case SUPPRESS:
        break;
    }
  }

  private void parseEvent(final Event event) throws Exception {
    final List<AnalyticsTracking> successfulProviders = new ArrayList<>();

    for (final WriteEvent writeEvent : event.writeEvents) {
      if (writeEvent != null && writeEvent.trackingResult.first) {
        successfulProviders.add(writeEvent.provider);
      } else {
        if (writeEvent != null) {
          handleError(writeEvent.trackingResult.second);
        }

        return;
      }
    }

    write(event.description, successfulProviders);
  }

  /**
   * Records the event.
   *
   * @param event The event to record.
   * @throws Exception The exception thrown when ErrorHandlingOption.STRICT is used.
   */
  void track(final Event event) throws Exception {
    events.add(event);
    parseEvent(event);
  }

  private void write(final String description, final List<AnalyticsTracking> providers) {
    switch (outputOption) {
      case NONE:
        return;
      case VERBOSE:
        for (final AnalyticsTracking provider : providers) {
          writeOut(provider.providerName(), description);
        }
        break;
      case SIMPLE:
        writeOut(ANALYTICS_PROVIDER_NAME, description);
        break;
    }
  }

  private void writeOut(final String name, final String message) {
    for (final Output source : outputSources) {
      source.print(String.format(WRITE_OUT_FORMAT, name, message));
    }
  }
}

File: simcoe/src/main/java/com/prolificinteractive/simcoe/Tracker.kt
package com.prolificinteractive.simcoe

import com.prolificinteractive.simcoe.trackers.AnalyticsTracking
import java.util.ArrayList

/**
 * A recorder for SimcoeEvents.
 * @param outputOption The output options.
 * @param errorOption  The error options.
 */
class Tracker @JvmOverloads constructor(
    var outputOption: OutputOptions = OutputOptions.VERBOSE,
    var errorOption: ErrorHandlingOption = ErrorHandlingOption.DEFAULT,
    vararg outputs: Output) {

  constructor() : this(outputs = ConsoleOutput())

  private val outputSources = outputs

  companion object {
    const private val ANALYTICS_PROVIDER_NAME = "Analytics"
    const private val ANALYTICS_ERROR = "** ANALYTICS ERROR **"
  }

  /**
   * Records the event.
   *
   * @param event The event to record.
   * @throws Exception The exception thrown when ErrorHandlingOption.STRICT is used.
   */
  internal fun track(event: Event) {
    val successfulProviders = ArrayList<AnalyticsTracking>()

    for (writeEvent in event.writeEvents) {
      if (!writeEvent.trackingResult.isError) {
        successfulProviders.add(writeEvent.provider)
      } else {
        handleError(writeEvent.provider.providerName(), writeEvent.trackingResult.errorMessage)
        return
      }
    }

    write(event.description, successfulProviders)
  }

  @Throws(SimcoeException::class)
  private fun handleError(name: String, errorMessage: String?) {
    when (errorOption) {
      ErrorHandlingOption.DEFAULT -> {
        val errorFormatted = String.format("%s => %s",
            name + (errorMessage ?: "Unknown errorMessage"))
        outputSources.writeOut(ANALYTICS_ERROR, errorFormatted)
      }
      ErrorHandlingOption.SUPPRESS -> {
      }
      ErrorHandlingOption.STRICT -> throw SimcoeException(errorMessage)
    }
  }

  private fun write(description: String, providers: List<AnalyticsTracking>) {
    when (outputOption) {
      OutputOptions.NONE -> return
      OutputOptions.VERBOSE -> providers.forEach {
        outputSources.writeOut(it.providerName(),
            description)
      }
      OutputOptions.SIMPLE -> outputSources.writeOut(ANALYTICS_PROVIDER_NAME, description)
    }
  }
}

